benches/nexmark/generator/bids.rs

```diff
@@ -0,0 +1,99 @@
+//! Generates bids for the Nexmark streaming data source.
+//!
+//! API based on the equivalent [Nexmark Flink PersonGenerator API](https://github.com/nexmark/nexmark/blob/v0.2.0/nexmark-flink/src/main/java/com/github/nexmark/flink/generator/model/BidGenerator.java).
+use super::strings::next_string;
+use cached::{proc_macro::cached, SizedCache};
+use rand::{thread_rng, Rng};
+
+const CHANNELS_NUMBER: usize = 10_000;
+
+const BASE_URL_PATH_LENGTH: usize = 5;
+
+// Similar to the Java implementation, the cached version of
+// `get_new_channel_instance` needs to create its own Rng, as it's not currently
+// possible to pass a generic Rng in while using the cached macro.
+#[cached(
+    type = "SizedCache<usize, (String, String)>",
+    create = "{ SizedCache::with_size(CHANNELS_NUMBER) }"
+)]
+fn get_new_channel_instance(channel_number: usize) -> (String, String) {
```

Comment:
Can't we just do this caching manually by storing a cache within the `NextmarkGenerator` struct, that way we cold provide an rng for it?

Reply:
By manually, do you mean without the `cached` crate all together? In which case, yes, though I think it's nice to avoid the extra code by just using a fn here? Or maybe you mean still using the `cached` library but manually instantiating a `SizedCache`... - hmm, that could be nicer - let me try.

Reply:
I've done the latter and it works nicely - thanks. One thing I did notice while doing this is that, since the use of the `rng` requires it to be `&mut`, every method that uses the `rng` field, needs `&mut self`, which in turn means the compiler has to assume (a ref to) the whole struct is borrowed mutably, rather than just the `rng` field. It's for this reason that I've not also moved the `get_base_url` function into the structs implementation, as currently I can just pass the one field as a mutable borrow. Let me know if I'm missing something here and I'll update it in another branch.

Reply:
Huh, though with my latest change it seems the compiler is failing to compile the `time` crate with
```
thread 'rustc' panicked at 'index out of bounds: the len is 1 but the index is 2', compiler\rustc_builtin_macros\src\format.rs:1001:28
```

Reply:
OK, it seems whatever issue that was causing `rustc` to panic in the nightly builds was now fixed :)